### PR TITLE
Paulhatcher/ent 5408

### PR DESCRIPTION
**Pull request by PaulHatcherR3**
Added support for synthetic methods with default arguments

### CODE REVIEW

File: quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/InstrumentClass.java [hunk 1]
```diff
@@ -277,6 +317,28 @@ public void visitEnd() {
         super.visitEnd();
     }
 
+    private boolean compareSuspendableArgs(Type[] expectedMethodTypes, MethodNode mns) {
+        final Type[] candidateMethodTypes = Type.getArgumentTypes(mns.desc);
+        for (int i = 0; i < candidateMethodTypes.length; i++) {
+            if (!Objects.equals(candidateMethodTypes[i], expectedMethodTypes[i + 1])) {
```
Comment (rick-r3) [inline quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/InstrumentClass.java:323]: I think it is worth a comment on this function illustrating what the signatures of the method signatures look like when they "match".  I had an impression in my head of what I thought they looked like which doesn't align with this, but then realised I don't really know.

e.g. 
`fun foo$default(arg0, arg1)`
`fun foo(<mask>, arg0, arg1)`

Comment (josephzunigadaly) [inline quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/InstrumentClass.java:323]: Do we need to check that `i + 1` is within range for the `expectedMethodTypes` array?

**Comment by rick-r3** [comment]
Does this solve the kotlin 1.3 problems ?

File: quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/InstrumentClass.java (hunk 1 above)
Comment (PaulHatcherR3) [reply quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/InstrumentClass.java]: Added additional comment

Comment (PaulHatcherR3) [reply quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/InstrumentClass.java]: Checked size of array before indexing

File: build.gradle [hunk 2]
```diff
@@ -51,7 +51,7 @@ subprojects {
     ext.quasarJar = "${rootProject.projectDir}/quasar-core/build/libs/quasar-core-${version}${ext.java8 ? "" : "-jdk7"}.jar" // project(':quasar-core').jar.archivePath
 
     ext.asmVer    = '5.2'
-    ext.kotlinVer = '1.1.3-2'
+    ext.kotlinVer = '1.3.70'
```
[the remark this answers is not in the document]
Comment (PaulHatcherR3) [reply build.gradle:54]: This is a version beyond 1.3.40 that exposes the problem.
I can change to 1.3.72 if required?
But was considering making this overidable on the command line so we can test different versions in Jenkins.

File: build.gradle [hunk 3]
```diff
@@ -166,7 +166,7 @@ subprojects {
         systemProperty 'co.paralleluniverse.monitoring.flightRecorderSize', '200000'
 
         if (project.name != 'quasar-core') {
-            jvmArgs "-javaagent:${quasarJar}" // =vdmc (verbose, debug, allow monitors, check class)
+            jvmArgs "-javaagent:${quasarJar}" // vdc (verbose, debug, allow monitors, check class)
```
[the remark this answers is not in the document]
Comment (PaulHatcherR3) [reply build.gradle:169]: I'll put it back to previous comment, I changed it after adding it to jvmArgs and did not restore it.

**Comment by PaulHatcherR3** [comment]
@rick-r3  With regard to Kotlin 1.3 problems this solves the specific issue of instrumenting methods with default arguments.

File: quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/InstrumentClass.java [hunk 4]
```diff
@@ -43,10 +43,14 @@
 
 import static co.paralleluniverse.fibers.instrument.Classes.*;
 import static co.paralleluniverse.fibers.instrument.QuasarInstrumentor.ASMAPI;
+
```
Comment (blsemo) [inline quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/InstrumentClass.java:46]: Might be just a nitpick, but why blank lines between import statements?

Comment (PaulHatcherR3) [reply quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/InstrumentClass.java:46]: Visual grouping, lines above are "import static" lines below are straight "import".

File: quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/InstrumentClass.java [hunk 5]
```diff
@@ -209,17 +225,25 @@ private void commit() {
...
-                            @Override
-                            public void visitEnd() {
-                                // don't call visitEnd on MV
-                            }
-                        }); // write method as-is
-                        this.mv = _mv;
+
+                        // If we may have a synthesized default method then save it for later processing.
+                        if (((access & Opcodes.ACC_STATIC) == Opcodes.ACC_STATIC) &&
+                            ((access & Opcodes.ACC_SYNTHETIC) == Opcodes.ACC_SYNTHETIC) &&
```
[the remark this answers is not in the document]
Comment (PaulHatcherR3) [reply quasar-core/src/main/java/co/paralleluniverse/fibers/instrument/InstrumentClass.java]: Modified as suggested.

File: quasar-kotlin/build.gradle [hunk 6]
```diff
@@ -16,7 +16,7 @@ dependencies {
     provided project(path: ':quasar-core', configuration: "jdk8Archives")
```
[the remark this answers is not in the document]
Comment (PaulHatcherR3) [reply quasar-kotlin/build.gradle]: Modified as suggested

**Pull request merged**